### PR TITLE
feat(parquet): omit unknown fields from writes

### PR DESCRIPTION
**Pull request by manuzhang**
## Which issue does this PR close?

No issue is currently linked.

## What changes are included in this PR?

This is the Parquet-write follow-up to #2773 and is intentionally opened as a draft.

- Builds the physical Parquet Arrow schema by omitting Unknown fields.
- Rejects containers that would have no valid physical Parquet representation.
- Projects incoming record batches to the physical schema before writing while preserving nested container structure and row counts.
- Updates NaN statistics traversal to work against the projected physical batch.

Until #2773 merges, this draft temporarily includes the Unknown logical-type commit in its cumulative diff. After #2773 merges, the branch will be rebased onto `main` so this PR contains only the Parquet-write commit.

## Are these changes tested?

- Added schema and batch-projection regressions for top-level and nested Unknown fields, invalid empty containers, and name-based matching.
- `cargo test -p iceberg --lib` (1,594 passed at this follow-up layer)
- `make check-fmt`
- `make check-clippy`
- `git diff --check`

## AI Disclosure

This PR was prepared with assistance from Codex.

### CODE REVIEW

**Comment by dannycjones** [comment]
Is this ready for review, after a rebase @manuzhang?

**Comment by manuzhang** [comment]
@dannycjones This is ready for review now, thanks!

File: crates/iceberg/src/arrow/schema.rs [hunk 1]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
...
+            };
+            if struct_type.fields().len() != arrow_fields.len() {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    format!(
+                        "Arrow and Iceberg struct field counts differ for field {}",
+                        field.id
+                    ),
+                ));
+            }
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs]: TODO: How does this look when something is a default?

File: crates/iceberg/src/arrow/schema.rs [hunk 2]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
...
+                })
+                .collect::<Result<Vec<_>>>()?;
+            if fields.is_empty() {
+                return Err(Error::new(
+                    ErrorKind::FeatureUnsupported,
+                    format!(
+                        "Cannot write struct field {} with no Parquet physical fields",
+                        field.id
+                    ),
+                ));
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs:838]: It's a weird edge case, but I think we should add a test where someone tries to write a Parquet file for an empty Iceberg schema - what do we expect to happen?

I'm also wondering how this will will change with #3250, but I think we can just leave it up to that workstream to tackle it and make any necessary changes.

File: crates/iceberg/src/arrow/schema.rs [hunk 3]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
     }
 }
 
+fn parquet_write_arrow_field(
+    field: &NestedFieldRef,
+    arrow_field: &FieldRef,
+) -> Result<Option<FieldRef>> {
+    let data_type = match field.field_type.as_ref() {
+        Type::Primitive(PrimitiveType::Unknown) => return Ok(None),
+        Type::Struct(struct_type) => {
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs:802]: For this match, I'd consider matching on both the Iceberg type and the Arrow type and having a more general error message.

```rust
let data_type = match (field.field_type.as_ref(), arrow_field.data_type()) {
    (Type::Primitive(PrimitiveType::Unknown), _) => return Ok(None),
    (Type::Struct(struct_type), DataType::Struct(arrow_fields)) => {
        // ...
    }
    // ...
    (Type::Primitive(_) | Type::Variant(_), arrow_type) => arrow_type.clone(),
    (iceberg_type, arrow_type) => {
        return Err(Error::new(
            ErrorKind::Unexpected,
            format!(
                "Unexpected Arrow type {} for Iceberg type {} with field ID {}",
                arrow_field.data_type(),
                iceberg_type,
                field.id,
            ),
        ));
    }
};
```

I don't think its worth the amount of boiler plate for a slightly richer error message here.

File: crates/iceberg/src/arrow/schema.rs [hunk 4]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
     }
 }
 
+fn parquet_write_arrow_field(
+    field: &NestedFieldRef,
+    arrow_field: &FieldRef,
+) -> Result<Option<FieldRef>> {
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs]: Can you add a Rustdoc to provide a guide to what this fn needs to do?

```suggestion
/// Map the Iceberg field to the equivalent Arrow field, descending into nested types.
///
/// For some Iceberg types, no Arrow type is returned.
/// [`Type::Unknown`] is one such example, which has no physical Parquet representation.
fn parquet_write_arrow_field(
    field: &NestedFieldRef,
    arrow_field: &FieldRef,
) -> Result<Option<FieldRef>> {
```

File: crates/iceberg/src/arrow/schema.rs [hunk 5]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
...
+    let data_type = match field.field_type.as_ref() {
+        Type::Primitive(PrimitiveType::Unknown) => return Ok(None),
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs:801]: nitpick: can you keep the primitive arms together?

i.e. don't have struct between them. I am happy that we pattern match on `Unknown` itself.

File: crates/iceberg/src/arrow/schema.rs [hunk 6]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
...
+            })?;
+            let entries_field = Arc::new(
+                entries_field
+                    .as_ref()
+                    .clone()
+                    .with_data_type(DataType::Struct(vec![key_field, value_field].into())),
+            );
+            DataType::Map(entries_field, *ordered)
+        }
+        Type::Primitive(_) | Type::Variant(_) => arrow_field.data_type().clone(),
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs:923]: I'd rather we have `Type::Variant(_)` return a feature unsupported error.


```suggestion
        Type::Primitive(_) => arrow_field.data_type().clone(),
        Type::Variant(_) => return Err(Error::new(
            ErrorKind::FeatureUnsupported,
            format!("Field {} has variant type, which is not yet implemented",
                field.id,
            ),
        )),
```

File: crates/iceberg/src/arrow/schema.rs [hunk 7]
```diff
@@ -793,6 +793,169 @@ pub fn schema_to_arrow_schema(schema: &Schema) -> Result<ArrowSchema> {
...
+    };
+
+    Ok(Some(Arc::new(
+        arrow_field.as_ref().clone().with_data_type(data_type),
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs]: Do we need `.as_ref().clone()`? Can we not clone directly?

File: crates/iceberg/src/arrow/schema.rs [hunk 8]
```diff
@@ -2221,6 +2384,149 @@ mod tests {
...
+            ])
+            .build()
+            .unwrap();
+
+        assert!(
+            schema_to_arrow_schema_for_parquet_write(&schema)
+                .unwrap_err()
+                .message()
+                .contains("struct field 2")
+        );
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs:2437]: can we expect_err and include more of the error message please?


```suggestion
        let err = schema_to_arrow_schema_for_parquet_write(&schema)
            .expect_err("conversion must fail when no fields are physically writable")
        assert_eq!(
            err.message(),
            "Cannot write struct field struct field 2 with no Parquet physical fields",
        );
```

similar for some of the tests following.

File: crates/iceberg/src/arrow/schema.rs [hunk 9]
```diff
@@ -2221,6 +2384,149 @@ mod tests {
...
+                    Type::List(ListType::new(
+                        NestedField::optional(
+                            2,
+                            "element",
+                            Type::Struct(StructType::new(vec![
+                                NestedField::optional(3, "unknown", PrimitiveType::Unknown.into())
+                                    .into(),
+                            ])),
+                        )
+                        .into(),
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/schema.rs:2493]: Please can you use `NestedField::list_element` to construct inner field.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 10]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+}
+
+fn fields_match(source: &FieldRef, target: &FieldRef, match_mode: FieldMatchMode) -> Result<bool> {
+    match match_mode {
+        FieldMatchMode::Id => {
+            Ok(get_field_id_from_metadata(source)? == get_field_id_from_metadata(target)?)
+        }
+        FieldMatchMode::Name => Ok(source.name() == target.name()),
+    }
+}
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:227]: Let's move this alongside `match_field` in `value.rs`.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 11]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+
+fn validate_source_field_for_parquet_write(
+    source: &FieldRef,
+    target: &FieldRef,
+    match_mode: FieldMatchMode,
+) -> Result<()> {
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:233]: Can you add Rustdoc for this one? What's the point of this fn?

Why for example might we skip the map values?

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 12]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+            || {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    format!(
+                        "Field {} is not present in the configured Iceberg schema for Parquet write",
+                        source_field.name()
+                    ),
+                )
+            },
+        )?;
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:292]: nitpick:


```suggestion
        let Some(target_index) = find_field_index(target_fields, source_field, match_mode)? else {
            return Err(Error::new(
                ErrorKind::DataInvalid,
                format!(
                    "Field {} is not present in the configured Iceberg schema for Parquet write",
                    source_field.name()
                ),
            ));
        };
```

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 13]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+            let values =
+                project_array_for_parquet_write(source.values(), target_element, match_mode)?;
+            Ok(Arc::new(ListArray::try_new(
+                target_element.clone(),
+                source.offsets().clone(),
+                values,
+                source.nulls().cloned(),
+            )?))
+        }
+        DataType::Map(target_entries, ordered) => {
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:379]: The name here is quite confusing - it doesn't portray that `target_entries` is the Arrow field for the entries in the map.


```suggestion
        DataType::Map(target_entries_field, ordered) => {
```

and use the similar naming on each arm?

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 14]
```diff
@@ -309,9 +584,18 @@ impl SchemaVisitor for IndexByParquetPathName {
...
+/// The input batches are matched to the configured Iceberg schema by field ID by default, or by
+/// name when configured through [`ParquetWriterBuilder::with_match_mode`]. In ID mode, every input
+/// field must contain valid Iceberg field-ID metadata. Fields with
+/// [`PrimitiveType::Unknown`] are accepted in logical input batches and omitted from the physical
+/// Parquet file.
 pub struct ParquetWriter {
     schema: SchemaRef,
+    logical_arrow_schema: ArrowSchemaRef,
+    arrow_schema: ArrowSchemaRef,
+    match_mode: FieldMatchMode,
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:598]: I think we can be clearer.


```suggestion
    /// Arrow schema converted from the Iceberg table schema.
    table_arrow_schema: ArrowSchemaRef,
    /// Arrow schema, where the equivalent table schema has been projected into what the new Parquet data file should contain.
    parquet_arrow_schema: ArrowSchemaRef,
    match_mode: FieldMatchMode,
```

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 15]
```diff
@@ -848,6 +1136,322 @@ mod tests {
...
+            FieldMatchMode::Name,
+        )
+        .unwrap();
+
+        let values = projected
+            .column(0)
+            .as_any()
+            .downcast_ref::<Int32Array>()
+            .unwrap();
+        assert_eq!(values.value(0), 10);
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:1231]: Worth asserting that there's only one column here

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 16]
```diff
@@ -848,6 +1136,322 @@ mod tests {
...
+    }
+
+    #[test]
+    fn test_project_batch_for_parquet_rejects_unconfigured_fields() {
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:1235]:
```suggestion
    fn test_project_batch_for_parquet_rejects_unexpected_fields() {
```

Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:1235]: I think this will break DataFusion's table provider, since it injects `_partition` and we don't strip it in the splitter. (It's not part of this repo anymore unfortunately, but I would recommend taking a look at the splitter and also checking out the DataFusion table provider to see if partitoned tests still work)

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 17]
```diff
@@ -848,6 +1136,322 @@ mod tests {
...
+
+    #[test]
+    fn test_project_batch_for_parquet_validates_list_element_identity() {
+        let schema = Schema::builder()
+            .with_fields(vec![
+                NestedField::optional(
+                    1,
+                    "list",
+                    Type::List(ListType::new(
+                        NestedField::optional(2, "element", PrimitiveType::Int.into()).into(),
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:1335]: please use `NestedField::list_element`

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 18]
```diff
@@ -848,6 +1136,322 @@ mod tests {
...
+                &batch,
+                logical_schema.clone(),
+                target_schema.clone(),
+                match_mode,
+            )
+            .unwrap_err();
+            assert!(
+                error.to_string().contains("List element field"),
+                "unexpected error: {error}"
+            );
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:1405]: Please assert on more of the error message here

File: crates/iceberg/src/arrow/nan_val_cnt_visitor.rs [hunk 19]
```diff
@@ -167,17 +241,8 @@ impl NanValueCountVisitor {
...
     pub fn compute(&mut self, schema: SchemaRef, batch: RecordBatch) -> Result<()> {
-        let arrow_arr_partner_accessor = ArrowArrayAccessor::new_with_match_mode(self.match_mode);
-
         let struct_arr = Arc::new(StructArray::from(batch)) as ArrayRef;
-        visit_struct_with_partner(
-            schema.as_struct(),
-            &struct_arr,
-            self,
-            &arrow_arr_partner_accessor,
-        )?;
```
Comment (dannycjones) [inline crates/iceberg/src/arrow/nan_val_cnt_visitor.rs]: We still leave this type implementing the visitor trait despite no longer using it. We should remove it if we really want to no longer use that trait.

File: crates/iceberg/src/arrow/nan_val_cnt_visitor.rs
Comment (dannycjones) [inline crates/iceberg/src/arrow/nan_val_cnt_visitor.rs:1]: I'm OK with using the much simpler visitor here rather than implementing the trait.

That said, we should remove the old implementation.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 20]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+
+fn project_batch_for_parquet_write(
+    batch: &RecordBatch,
+    logical_schema: ArrowSchemaRef,
+    target_schema: ArrowSchemaRef,
+    match_mode: FieldMatchMode,
+) -> Result<RecordBatch> {
+    if batch.schema_ref() == &target_schema {
+        return Ok(batch.clone());
+    }
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:431]: Should we validate the the original target schema?

Open question.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 21]
```diff
@@ -617,28 +901,32 @@ fn resolve_writer_properties(
...
 
         self.current_row_num += batch.num_rows();
 
-        let batch_c = batch.clone();
+        let batch = project_batch_for_parquet_write(
+            batch,
+            self.logical_arrow_schema.clone(),
+            self.arrow_schema.clone(),
+            self.match_mode,
+        )?;
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs]: I think we should move this before we increment the `current_row_num` - wdyt?

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 22]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+                (DataType::Struct(source_fields), DataType::Struct(target_fields)) => {
+                    validate_source_fields_for_parquet_write(
+                        source_fields,
+                        target_fields,
+                        match_mode,
+                    )
+                }
+                _ => Ok(()),
+            }
+        }
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:270]: I don't quite understand it deeply enough yet, but I think we're being oddly strict here and it may impact callers.

LLM raised that some callers may construct Arrow batches themselves, and when constructing lists then Arrow isn't going to attach fields IDs nor is it going to use `element` as the name of the field - it may use `item` instead.

I need to go back and review Arrow more.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 23]
```diff
@@ -617,28 +901,32 @@ fn resolve_writer_properties(
...
 
 impl FileWriter for ParquetWriter {
-    async fn write(&mut self, batch: &arrow_array::RecordBatch) -> Result<()> {
+    async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
         // Skip empty batch
         if batch.num_rows() == 0 {
             return Ok(());
         }
 
         self.current_row_num += batch.num_rows();
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:910]: Should this actually be after the `write`?

I'm not super comfortable with the behavior here, needs some more thinking.

What should `current_row_num` be given the different outcomes? IMO if we fail to project the batch, we didn't write anything so we shouldn't have incremented the row number yet.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 24]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+    if batch.schema_ref() == &target_schema {
+        return Ok(batch.clone());
+    }
+
+    let source_schema = batch.schema();
+    validate_source_fields_for_parquet_write(
+        source_schema.fields(),
+        logical_schema.fields(),
+        match_mode,
+    )?;
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:438]: This seems relatively expensive. This is occurring on every batch - I think it would be worth considering saving the result if, for example, we're being passed the same schema every time.

- If we already validated the target schema and its equal, sounds good.
- If we have seen this batch schema before and it was valid, then its still valid!

We're also doing a lot of walking of trees and comparing metadata.

Ultimately, I think there's a a risk this is a big source of slow down (especially on wide tables) and there are a bunch of optimizations we could reach for.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs (hunk 12 above)
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:292]: Each time here, we're iterating over the target schema which never changes. I think we should cache the column index to Iceberg field ID.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs [hunk 25]
```diff
@@ -184,6 +196,269 @@ impl FileWriterBuilder for ParquetWriterBuilder {
...
+                            target_field.name()
+                        ),
+                    )
+                })?;
+            project_array_for_parquet_write(batch.column(index), target_field, match_mode)
+        })
+        .collect::<Result<Vec<_>>>()?;
+    let options = RecordBatchOptions::default()
+        .with_match_field_names(false)
+        .with_row_count(Some(batch.num_rows()));
```
Comment (dannycjones) [inline crates/iceberg/src/writer/file_writer/parquet_writer.rs:458]: I'm not convinced we should turn off `match_field_names`. Is it expensive?

If you think its a good idea, can you add a comment to justify?

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs (hunk 24 above)
Comment (dannycjones) [reply crates/iceberg/src/writer/file_writer/parquet_writer.rs:438]: I've been playing a bit with an LLM, it suggests something like this after some back and forth.

The idea is to do iterate over the fields rather than trying to traverse, in order to build lists you can then `zip` with.

```rust
/// Correspondence between a batch's fields and the schema's fields at one level.
struct FieldAlignment {
    /// For each target field, the source index feeding it. None = absent from the batch.
    target_to_source: Vec<Option<usize>>,
    /// Source indices matching no target field.
    unmatched_sources: Vec<usize>,
}

fn align_fields(
    source: &Fields,
    target: &Fields,
    match_mode: FieldMatchMode,
) -> Result<FieldAlignment> {
    #[derive(Hash, PartialEq, Eq)]
    enum FieldKey<'a> {
        Id(i32),
        Name(&'a str),
    }

    let mut by_key = HashMap::with_capacity(source.len());
    for (index, field) in source.iter().enumerate() {
        let key = match match_mode {
            FieldMatchMode::Id => FieldKey::Id(get_field_id_from_metadata(field)?),
            FieldMatchMode::Name => FieldKey::Name(field.name()),
        };
        if by_key.insert(key, index).is_some() {
            return Err(Error::new(ErrorKind::DataInvalid, format!(
                "Multiple record batch fields match configured field {}", field.name())));
        }
    }

    let mut matched = vec![false; source.len()];
    let mut target_to_source = Vec::with_capacity(target.len());
    for target_field in target {
        let key = match match_mode {
            FieldMatchMode::Id => FieldKey::Id(get_field_id_from_metadata(target_field)?),
            FieldMatchMode::Name => FieldKey::Name(target_field.name()),
        };
        let found = by_key.get(&key).copied();
        if let Some(index) = found {
            matched[index] = true;
        }
        target_to_source.push(found);
    }

    Ok(FieldAlignment {
        target_to_source,
        unmatched_sources: (0..source.len()).filter(|i| !matched[*i]).collect(),
    })
}
```

Obviously, this optimization is assuming that validating is the right thing to do.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs (hunk 10 above)
Comment (dannycjones) [reply crates/iceberg/src/writer/file_writer/parquet_writer.rs:227]: If we implement some of the optimizations mention in other threads, this function may no longer be needed.

**Review by dannycjones — COMMENTED** [review]
Thanks for working on this, @manuzhang! I have quite a few comments, I've been working on similar code myself when trying to implement both geo and `write-default` so I've been stumbling around here a bunch. I'm also still finding my feet around Arrow and how its used in iceberg-rust.

File: crates/iceberg/src/writer/file_writer/parquet_writer.rs (hunk 12 above)
Comment (dannycjones) [reply crates/iceberg/src/writer/file_writer/parquet_writer.rs:292]: Maybe related: https://github.com/apache/iceberg-rust/pull/3072#discussion_r4088427820
